parlai/agents/image_seq2seq/modules.py

```diff
@@ -0,0 +1,201 @@
+# LICENSE file in the root directory of this source tree.
+"""
+Modules for ImageSeq2seqAgent Agent.
+"""
+import torch
+import torch.nn as nn
+from typing import List, Tuple
+
+from parlai.agents.transformer.modules import (
+    TransformerGeneratorModel,
+    TransformerEncoder,
+)
+from parlai.core.dict import DictionaryAgent
+from parlai.core.opt import Opt
+
+
+class ImageSeq2seqModel(TransformerGeneratorModel):
+    """
+    ImageSeq2seqModel.
+
+    Just TGA that can encode image with encoder.
+    """
+
+    def __init__(self, opt: Opt, dictionary: DictionaryAgent):
+        if opt.get('n_positions'):
+            # if the number of positions is explicitly provided, use that
+            n_positions = opt['n_positions']
+        else:
+            # else, use the worst case from truncate
+            n_positions = max(
+                opt.get('truncate') or 0,
+                opt.get('text_truncate') or 0,
+                opt.get('label_truncate') or 0,
+            )
+            if n_positions == 0:
+                # default to 1024
+                n_positions = 1024
+
+        super().__init__(opt, dictionary)
+        self.encoder = ContextWithImageEncoder(
+            n_heads=opt['n_heads'],
+            n_layers=opt['n_layers'],
+            embedding_size=opt['embedding_size'],
+            ffn_size=opt['ffn_size'],
+            vocabulary_size=len(dictionary),
+            embedding=self.embeddings,
+            dropout=opt['dropout'],
+            attention_dropout=opt['attention_dropout'],
+            relu_dropout=opt['relu_dropout'],
+            padding_idx=self.pad_idx,
+            learn_positional_embeddings=opt['learn_positional_embeddings'],
+            embeddings_scale=opt['embeddings_scale'],
+            reduction_type=None,
+            n_positions=n_positions,
+            n_segments=opt.get('n_segments', 0),
+            activation=opt['activation'],
+            variant=opt['variant'],
+            output_scaling=opt['output_scaling'],
+            image_encoder_num_layers=opt['image_encoder_num_layers'],
+            image_features_dim=opt['image_features_dim'],
+            use_cuda=not opt['no_cuda'] and torch.cuda.is_available(),
+        )
+
+
+class ContextWithImageEncoder(TransformerEncoder):
+    """
+    ContextWithImage Module.
+
+    Encodes image and context via simple concatenation.
+    """
+
+    def __init__(self, *args, **kwargs):
+        """
+        Override TransformerEncoder __init__.
+
+        Setup the image encoder; create some dummy tensors for inserting image into
+        input
+        """
+        self.n_img_layers = kwargs.pop('image_encoder_num_layers')
+        self.img_dim = kwargs.pop('image_features_dim')
+        self.use_cuda = kwargs.pop('use_cuda')
+        super().__init__(*args, **kwargs)
+        self._build_image_encoder()
+        self.dummy_image_enc = torch.zeros((self.embedding_size))
+        self.ones_mask = torch.ones(1).bool()
+        if self.use_cuda:
```

Comment:
so it's better to avoid passing around a `use_cuda` or `device` flag in model code. Instead, create the tensor and use `tensor.to(known_tensor.device)`. For example here, you might use self.somelayer.device